### ISSUE TITLE
Code first registered interfaces cause errors if out of order.

### ISSUE THREAD

**Issue by DanielMillerSEL** [graphql-dotnet/graphql-dotnet#1046]
## Description
When doing code first, the registration order of interfaces and graph types can cause run time errors when trying to serialize an object that implements an interface. The errors always seem to relate to receiving null for the graph parent type, instead of the concrete instance.

### Steps to reproduce

The error occurs when you use the code first approach and register the interface graph type before registering an implementer of the interface.

Here is one of the smallest examples I have where the error occurs:
```
using GraphQL;
using GraphQL.Types;

namespace GraphQLInterfaceErrorTest
{
  class Program
  {
    static void Main()
    {
      var schema = new Schema { Query = new QueryGraphType() };
      // Comment this line out, or reverse it and ImplementationGraphType and it works as expected.
      schema.RegisterType<InterfaceGraphType>();
      schema.RegisterType<ImplementationGraphType>();
      var response = schema.Execute(_ =>
      {
        _.Query = @"
query {
  inst {
    id
  }
}
";
      });
      // expect response does not have errors.
     }
  }

  public class QueryGraphType : ObjectGraphType
  {
    public QueryGraphType()
    {
      Field<InterfaceGraphType>("inst", resolve: c => new Implementation());
    }
  }

  public interface IInterface
  {
    string Id { get; }
  }
  public class InterfaceGraphType : InterfaceGraphType<IInterface>
  {
    public InterfaceGraphType()
    {
      Field(i => i.Id);
    }
  }

  public class Implementation : IInterface
  {
    public string Id => "Data!";
  }

  public class ImplementationGraphType : ObjectGraphType<Implementation>
  {
    public ImplementationGraphType()
    {
      Field(i => i.Id);
      Interface<InterfaceGraphType>();
    }
  }
}
```

### Expected result

```
{
  "data": {
    "inst": {
      "id": "Data!"
    }
  }
}
```

### Actual result

```
{
  "data": {
    "inst": null
  },
  "errors": [
    {
      "message": "Abstract type InterfaceGraphType must resolve to an Object type at runtime for field QueryGraphType.inst with value 'GraphQLInterfaceErrorTest.Implementation', received 'null'.",
      "locations": [
        {
          "line": 3,
          "column": 3
        }
      ],
      "path": [
        "inst"
      ]
    }
  ]
}
```

### Workaround

If you swap the order of these two lines, or remove the first one, no error occurs.
```
schema.RegisterType<InstanceGraphType>();
schema.RegisterType<ImplementationGraphType>();
```
Workarounds are easy, but the bug is not straightforward to figure out based on the error message provided.

### Environment

GraphQL 2.4.0

.NET Core 3.0
.NET Core 2.2
.NET Core 2.1
Happens on different systems and platforms including windows 10 and the asp.net core container, the error is reliably repeatable.

**Comment by joemcbride** [comment]
You should never need to call RegisterType with an interface type.  What has lead you to that solution?  Also, this framework provides a class called InterfaceGraphType already, so are you sure this code compiles and is using the expected types?

**Comment by joemcbride** [comment]
Also note - you’re not setting any names on your GraphTypes.  What does your Schema look like when you print it with SchemaPrinter?

**Comment by DanielMillerSEL** [comment]
This code example is just to highlight the issue. In our actual application names do not collide and we do manually provide the name in our graph types. This code does compile though, so I do not believe it is running into the name conflict you describe.
We kept running into issues where it was more maintainable to do schema first, so we wrote a code generator that would produce the graphql-dotnet binding classes from a schema. We use some systems using resolvers that are not supported out of the box, such as using the asp dependency injector that made normal gql.net schema first undesirable.
We did the brute force approach with our codegen and registered every type listed in the schema, including interfaces. We already fixed it on our end by removing the interface registration, but it still feelt like a bug, hence the bug report.
Feel free to report this as closed and as working-as-intended, my goal is to report that its an unintended behavior, but I agree that it's from an odd usage that was an issue on our end. We just didn't know it was an issue until the error occurred, which was around the third interface we added to the schema, the first two worked as intended as one of their implementors were registered first.

PS: Without names declared, the schema printer prints the names long form as QueryGraphType, ImplementationGraphType , and InterfaceGraphType.
Here is the printed schema
```
schema {
  query: QueryGraphType
}

# The `Date` scalar type represents a year, month and day in accordance with the
# [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) standard.
scalar Date

# The `DateTime` scalar type represents a date and time. `DateTime` expects
# timestamps to be formatted in accordance with the
# [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) standard.
scalar DateTime

# The `DateTimeOffset` scalar type represents a date, time and offset from UTC.
# `DateTimeOffset` expects timestamps to be formatted in accordance with the
# [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) standard.
scalar DateTimeOffset

scalar Decimal

type ImplementationGraphType implements InterfaceGraphType {
  id: String!
}

interface InterfaceGraphType {
  id: String!
}

# The `Milliseconds` scalar type represents a period of time represented as the total number of milliseconds.
scalar Milliseconds

type QueryGraphType {
  inst: InterfaceGraphType
}

# The `Seconds` scalar type represents a period of time represented as the total number of seconds.
scalar Seconds
```

**Comment by sungam3r** [comment]
@DanielMillerSEL Thanks fro the repro. Fixed in #2314 and will be published in v4.
